plugin.video.invidious/resources/lib/invidious_plugin.py

```diff
@@ -0,0 +1,338 @@
+    query top."""
+
+    def __init__(self, history_path, depth=10):
+        self.history_path = history_path
+        self.depth = depth
+
+
+    def push(self, query):
+        if xbmcvfs.exists(self.history_path):
+            with open(self.history_path, "r") as file:
+                queries = json.load(file)
+        else:
+            queries = []
+
+        if query in queries:
+            # Remove existing entry to move it forward
+            queries.remove(query)
+
+        queries.insert(0, query)
+
+        queries = queries[:self.depth]
+
+        with open(self.history_path, "w+") as file:
+            json.dump(queries, file)
+
+
+    def queries(self):
+        if not xbmcvfs.exists(self.history_path):
+            return []
+        with open(self.history_path, "r") as file:
+            return json.load(file)
+
+
+class InvidiousPlugin:
+    # special lists provided by the Invidious API
+    SPECIAL_LISTS = ("trending", "popular")
+
+    INSTANCESURL = "https://api.invidious.io/instances.json?sort_by=type,health"
+    def __init__(self, base_url, addon_handle, args):
+        self.base_url = base_url
+        self.addon_handle = addon_handle
+        self.addon = xbmcaddon.Addon()
+        self.args = args
+        path = xbmcvfs.translatePath(self.addon.getAddonInfo('profile'))
+        self.search_history = SearchHistory(path + 'search-history.json', 20)
+
+        instance_url = xbmcplugin.getSetting(self.addon_handle, "instance_url")
+        if 'auto' == instance_url:
+            instance_url = self.instance_autodetect()
+        xbmc.log(f'invidous using instance {instance_url}.', xbmc.LOGINFO)
+        self.api_client = invidious_api.InvidiousAPIClient(instance_url)
+        self.disable_dash = \
+            ("true" == xbmcplugin.getSetting(self.addon_handle,
+                                             "disable_dash"))
+    def instance_autodetect(self):
+        xbmc.log('invidious picking instance automatically.', xbmc.LOGINFO)
+
+        response = requests.get(self.INSTANCESURL, timeout=5)
+        data = response.json()
+        for instanceinfo in data:
+            xbmc.log('invidious considering instance ' + str(instanceinfo), xbmc.LOGDEBUG)
+            instancename, instance = instanceinfo
+            if 'https' == instance['type']:
+                instance_url = instance['uri']
+                # Make sure the instance work for us.  This test avoid
+                # those rejecting us with HTTP status 429.
+                api_client = invidious_api.InvidiousAPIClient(instance_url)
+                if api_client.fetch_special_list(self.SPECIAL_LISTS[0]):
+                    return instance_url
+
+        xbmc.log('invidious no working https type instance returned from api.invidious.io.', xbmc.LOGWARNING)
+        # FIXME figure out how to show failing autodetection to the user.
+        dialog = xbmcgui.Dialog()
+        dialog.notification(
```

Comment:
In the future please move all "Notifications" and "Dialog" messages to localized language. THX

Reply:
Done upstream.